File: server/shared/dotnet/caching-local-default/LocalCacheTelemetry.cs
// -----------------------------------------------------------------------
// <copyright file="LocalCacheTelemetry.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.Caching.Local.Default;

using System.Diagnostics.Metrics;

/// <summary>
/// Static OTel <see cref="Meter"/> for the default local-cache impl.
/// Per-call observability is intentionally minimal — no spans, no per-op
/// logs — because the cache work itself is ~60ns and per-call instrumentation
/// would dominate it. Aggregate counters give us the signal that matters
/// (hit rate, miss rate, eviction count) without per-op overhead.
/// </summary>
internal static class LocalCacheTelemetry
{
    /// <summary>OTel meter for the default local cache.</summary>
    internal static readonly Meter SR_Meter = new("D2.Shared.Caching.Local", "1.0.0");

    /// <summary>Counter incremented on every cache hit.</summary>
    internal static readonly Counter<long> SR_Hits =
        SR_Meter.CreateCounter<long>("d2.cache.local.hits", description: "Local cache hits.");

    /// <summary>Counter incremented on every cache miss.</summary>
    internal static readonly Counter<long> SR_Misses =
        SR_Meter.CreateCounter<long>("d2.cache.local.misses", description: "Local cache misses.");

    /// <summary>Counter incremented on every cache write.</summary>
    internal static readonly Counter<long> SR_Sets =
        SR_Meter.CreateCounter<long>("d2.cache.local.sets", description: "Local cache writes.");

    /// <summary>Counter incremented on every explicit removal.</summary>
    internal static readonly Counter<long> SR_Removes = SR_Meter.CreateCounter<long>(
        "d2.cache.local.removes",
        description: "Local cache removals (explicit).");

    /// <summary>Counter incremented on every implicit eviction (capacity / TTL).</summary>
    internal static readonly Counter<long> SR_Evictions = SR_Meter.CreateCounter<long>(
        "d2.cache.local.evictions",
        description: "Entries evicted by capacity / expiration.");
}

File: server/shared/typescript/caching/local-default/src/local-cache-telemetry.ts
// -----------------------------------------------------------------------
// Copyright (c) DCSV. All rights reserved.
// -----------------------------------------------------------------------

import { type Counter, metrics } from "@opentelemetry/api";

/**
 * OpenTelemetry meter name for the default local-cache implementation.
 * Hosts register this via MeterProvider setup so local-cache counters
 * reach OTLP / Prometheus exporters.
 */
// Byte-identical twin of .NET LocalCacheTelemetry: meter "D2.Shared.Caching.Local" v1.0.0,
// counters d2.cache.local.{hits,misses,sets,removes,evictions} with the same units and
// descriptions. Aggregate counters only - no tags, no spans, no logs (matches the .NET meter).
export const LOCAL_CACHE_METER_NAME = "D2.Shared.Caching.Local";

const LOCAL_CACHE_METER_VERSION = "1.0.0";

/** Bundle of the five aggregate counters owned by {@link DefaultLocalCache}. */
export interface LocalCacheCounters {
  hits: Counter;
  misses: Counter;
  sets: Counter;
  removes: Counter;
  evictions: Counter;
}

/**
 * Creates the five local-cache counters against the current global
 * MeterProvider. Call from the {@link DefaultLocalCache} constructor so
 * counters bind after telemetry bootstrap (module-load bind permanently
 * no-ops when imported before setup).
 */
export function createLocalCacheCounters(): LocalCacheCounters {
  const meter = metrics.getMeter(
    LOCAL_CACHE_METER_NAME,
    LOCAL_CACHE_METER_VERSION,
  );

  return {
    hits: meter.createCounter("d2.cache.local.hits", {
      unit: "{hit}",
      description: "Local cache hits.",
    }),
    misses: meter.createCounter("d2.cache.local.misses", {
      unit: "{miss}",
      description: "Local cache misses.",
    }),
    sets: meter.createCounter("d2.cache.local.sets", {
      unit: "{write}",
      description: "Local cache writes.",
    }),
    removes: meter.createCounter("d2.cache.local.removes", {
      unit: "{removal}",
      description: "Local cache removals (explicit).",
    }),
    evictions: meter.createCounter("d2.cache.local.evictions", {
      unit: "{eviction}",
      description: "Entries evicted by capacity / expiration.",
    }),
  };
}
